File: src/comms/CborGzipEncoder.py
from typing import Type
import gzip

from .CborEncoder import CborEncoder
from .jsonRpc import Request, Response, RequestOrResponse

class CborGzipEncoder(CborEncoder):

    def name(self) -> str:
        return 'cbor+gzip'

    def decode(self, message: bytes, cls: Type[RequestOrResponse]) -> RequestOrResponse:
        return CborEncoder.decode(self, gzip.decompress(message), cls)

    def encode(self, obj: RequestOrResponse) -> bytes:
        return gzip.compress(CborEncoder.encode(self, obj))

File: js/src/comms/CborGzipEncoder.ts
import pako from 'pako'

import CborEncoder from './CborEncoder'
import JsonRpcRequest from './JsonRpcRequest'
import JsonRpcResponse from './JsonRpcResponse'

/**
 * A [CBOR](http://cbor.io) + gzip message encoder
 */
export default class CborGzipEncoder extends CborEncoder {

  name (): string {
    return 'cbor+gzip'
  }

  encode (instance: JsonRpcRequest | JsonRpcResponse): Uint8Array {
    return pako.deflate(super.encode(instance))
  }

  decode<Type = JsonRpcRequest | JsonRpcResponse> (message: Uint8Array, constructor: new () => Type): Type {
    return super.decode(pako.inflate(message), constructor)
  }
}
